crates/core_simd/src/array.rs

```diff
@@ -0,0 +1,165 @@
+use crate::intrinsics;
+use crate::masks::*;
+use crate::vector::ptr::SimdConst;
+use crate::vector::*;
+
+/// A representation of a vector as an "array" with indices,
+/// implementing operations applicable to any vector type and
+/// describing relationships between vector and scalar types.
+pub trait SimdArray<const LANES: usize>: crate::LanesAtMost32
```

Comment:
My operating theory is that "array" generalizes here in a few important ways:

0. By default: arrays go on the stack, Vec goes to the heap. The name-clash with Vector might suggest the wrong things about data locality, which I've noticed Rustaceans care about a lot.
1. Assuming people successfully "deconflict" Vec and Vector, arrays are an existing type, so "this is an array, but SIMD" is only one bit of new data to absorb whereas "this is a vector, and SIMD" is two bits (three if counting the disambiguation bit).
2. The type is in fact more or less an array in the sense of readily admitting conversions to such (to the point of allowing some safe transmutes!), even if back again is slightly trickier.
3. People who use APL derivatives, Fortran, or ndarray libs seem to have no problem calling them arrays, even if their code  does vectorize quite nicely, and while we're from a slightly different "tradition", programming-wise, I think it's useful as a signpost that it's fine to embrace a level of abstraction from the vector registers... even if we still have to talk about them anyways, IMO.

But I'm not like... married to this? I just felt like experimenting with this alternative would be useful, especially after noticing that in conversations with other programmers I have to say "vectors... n-no, not Vec..." Even formal papers on portable SIMD APIs for C++ or Java have to add that caveat.

Reply:
I call simd values "registers", which is slightly wrong also, but not too much so. Maybe that applies more to `safe_arch` than stdsimd

Reply:
At that point wouldn't we want to strike "vector" from our terminology everywhere?

My concern is that as unfortunate the name clash is, we probably should still call them vectors simply because that's what they're called everywhere else.

Reply:
I think it's only confusing when you say it out loud and also follow the "it's spelled Vec but pronounced Vector" line. Everyone knows that you should spell it Vec and *also* pronounce it Vec.

In text form you wouldn't even have a problem.

Though I'll say that SimdArray doesn't really tell me what ops it supports. Can we just have these ops be on two separate traits called SimdScatter and SimdGather?

Reply:
Hmm. My thought is that having a single trait that defines the operations on the SIMD-vector-qua-SIMD-vector (or SIMD-array-qua-SIMD-array) will help understanding the API a fair amount. But I think we should discuss this while sync.

Reply:
> In text form you wouldn't even have a problem.
Unfortunately, this has not been my experience.

Reply:
> Can we just have these ops be on two separate traits called SimdScatter and SimdGather?

I think this line of reasoning leads to dozens of traits, and I'm fairly strongly opposed to following that route. I think the stdlib has been fairly conservative about trait definitions, which is a *good* thing.

Reply:
Ehhh, core::ops has one trait per op and it works fine.

Reply:
I think it would be fair to say that each op trait inherently needs to be separate since the ops are separate, this is a little more like Iterator, lots of functionality on one type.